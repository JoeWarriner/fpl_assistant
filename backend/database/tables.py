
from sqlalchemy import Column, Integer, Boolean, String, TIMESTAMP, ForeignKey,  UniqueConstraint, Float
from sqlalchemy.orm import relationship, DeclarativeBase

class Base(DeclarativeBase):
    pass

class Season(Base):
    __tablename__ = 'seasons'
    __table_args__ = (UniqueConstraint('start_year',  name='seasons__prevent_duplicate_import'),)

    id = Column(Integer, primary_key=True)
    is_current = Column(Boolean)
    start_year = Column(Integer)
    season = Column(String(35))
    
    team_seasons = relationship("TeamSeason", back_populates="season")
    gameweeks = relationship("Gameweek", back_populates="season")
    player_seasons = relationship("PlayerSeason", back_populates="season")
    fixtures = relationship("Fixture", back_populates='season')


class Team(Base):
    __tablename__ = 'teams'
    __table_args__ = (UniqueConstraint('fpl_id',  name='teams__prevent_duplicate_import'),)


    id = Column(Integer, primary_key=True)
    fpl_id = Column(Integer)
    team_name = Column(String)
    short_name = Column(String)
    
    
    team_seasons = relationship("TeamSeason", back_populates="team")

    

class TeamSeason(Base):
    __tablename__ = 'team_seasons'
    __table_args__ = (UniqueConstraint('fpl_id', 'season_id', name='team_seasons__prevent_duplicate_import'),)

    id = Column(Integer, primary_key=True)
    fpl_id = Column(Integer)
    team_id = Column(Integer, ForeignKey('teams.id'))
    season_id = Column(Integer, ForeignKey('seasons.id'))
    
    team = relationship("Team", back_populates="team_seasons")
    season = relationship("Season", back_populates="team_seasons")


class Gameweek(Base):
    __tablename__ = 'gameweeks'
    __table_args__ = (UniqueConstraint('gw_number', 'season_id', name='gameweeks__prevent_duplicate_import'), )
    
    id = Column(Integer, primary_key=True)
    deadline_time = Column(TIMESTAMP)
    finished = Column(Boolean)
    is_previous = Column(Boolean)
    is_next = Column(Boolean)
    is_current = Column(Boolean)
    gw_number = Column(Integer, nullable=False)
    season_id = Column(Integer, ForeignKey('seasons.id'), nullable= False)
    
    season = relationship("Season", back_populates="gameweeks")
    fixtures = relationship("Fixture", back_populates="gameweek")

class Fixture(Base):
    __tablename__ = 'fixtures'
    __table_args__ = (UniqueConstraint('fpl_code', name='fixtures__prevent_duplicate_import'), )

    
    id = Column(Integer, primary_key=True)
    gameweek_id = Column(Integer, ForeignKey('gameweeks.id'))
    away_team_id = Column(Integer, ForeignKey('teams.id'))
    home_team_id = Column(Integer, ForeignKey('teams.id'))
    away_team_difficulty = Column(Integer)
    home_team_difficulty = Column(Integer)
    season_id = Column(Integer, ForeignKey('seasons.id'))
    away_team_score = Column(Integer)
    home_team_score = Column(Integer)
    fpl_id = Column(Integer)
    fpl_code = Column(Integer, unique=True)
    kickoff_time = Column(TIMESTAMP)
    finished = Column(Boolean)
    started = Column(Boolean)
    
    gameweek = relationship("Gameweek", back_populates="fixtures")
    away_team = relationship("Team", foreign_keys=away_team_id, backref="away_fixtures")
    home_team = relationship("Team", foreign_keys=home_team_id, backref="home_fixtures")

    season = relationship("Season", back_populates="fixtures")

class PlayerPerformance(Base):
    __tablename__ = 'player_performances'
    __table_args__ = (
        UniqueConstraint('fixture_id', 'player_id', name = 'player_performances__prevent_duplicate_import'),
    )

    id = Column(Integer, primary_key=True)
    fixture_id = Column(Integer, ForeignKey('fixtures.id'))
    player_id = Column(Integer, ForeignKey('players.id'))
    team_id = Column(Integer, ForeignKey('teams.id'))
    opposition_id = Column(Integer, ForeignKey('teams.id'))
    minutes_played = Column(Integer)
    penalties_missed = Column(Integer)
    penalties_saved = Column(Integer)
    player_value = Column(Integer)
    red_cards = Column(Integer)
    yellow_cards = Column(Integer)
    selected = Column(Integer)
    total_points = Column(Integer)
    goals_scored = Column(Integer)
    goals_conceded = Column(Integer)
    clean_sheet = Column(Boolean)
    bonus = Column(Integer)
    assists = Column(Integer)
    was_home = Column(Boolean)
    difficulty = Column(Integer)
    saves = Column(Integer)
    expected_goals = Column(Float)
    expected_assists = Column(Float)
    expected_goals_conceded = Column(Float)
    influence = Column(Float)
    creativity =  Column(Float)
    threat = Column(Float)
    bps = Column(Integer)
    


    
    fixture = relationship("Fixture", backref="player_performances")
    player = relationship("Player", back_populates="player_performances")

    team = relationship("Team", foreign_keys=team_id, backref="player_performances")
    opposition = relationship("Team", foreign_keys=opposition_id, backref="opposition_player_performances")

class PlayerFixture(Base):
    __tablename__ = 'player_fixtures'
    __table_args__ = (
        UniqueConstraint('fixture_id', 'player_id', name = 'player_fixtures__prevent_duplicate_import'),
    )
    id = Column(Integer, primary_key=True)
    fixture_id = Column(Integer, ForeignKey('fixtures.id'))
    player_id = Column(Integer, ForeignKey('players.id'))
    team_id = Column(Integer, ForeignKey('teams.id'))
    opposition_id = Column(Integer, ForeignKey('teams.id')) 
    predicted_score = Column(Float, nullable=True)
    difficulty = Column(Integer)
    is_home = Column(Boolean)

    team = relationship('Team', foreign_keys=team_id, backref="team_player_fixtures")
    opposition = relationship('Team', foreign_keys=opposition_id, backref='opposition_player_fixtures')


    fixture = relationship("Fixture", backref="player_fixtures")
    player = relationship("Player", back_populates="player_fixtures")



class PlayerSeason(Base):
    __tablename__ = 'player_seasons'
    __table_args__ = (
        UniqueConstraint('fpl_id', 'season_id', name = 'player_seasons__prevent_duplicate_import'),
    )
    
    id = Column(Integer, primary_key=True)
    fpl_id = Column(Integer)
    player_id = Column(Integer, ForeignKey('players.id'))
    season_id = Column(Integer, ForeignKey('seasons.id'))
    position_id = Column(Integer, ForeignKey('positions.id'))
    
    player = relationship("Player", back_populates="player_seasons")
    season = relationship("Season", back_populates="player_seasons")
    position = relationship("Position", back_populates="player_seasons")

class Player(Base):
    __tablename__ = 'players'

    __table_args__ = (
        UniqueConstraint('fpl_id', name = 'players__prevent_duplicate_import'),
    )
    
    id = Column(Integer, primary_key=True)
    fpl_id = Column(Integer, unique=True)
    first_name = Column(String(35))
    second_name = Column(String(35))
    current_value = Column(Integer, nullable=True)
    
    player_performances = relationship("PlayerPerformance", back_populates="player")
    player_fixtures = relationship("PlayerFixture", back_populates="player")
    player_seasons = relationship("PlayerSeason", back_populates="player")

class Position(Base):
    __tablename__ = 'positions'
    __table_args__ = (
        UniqueConstraint('fpl_id', name = 'positions__prevent_duplicate_import'),
    )
    
    id = Column(Integer, primary_key=True)
    fpl_id = Column(Integer, unique=True)
    pos_name = Column(String(35))
    short_name = Column(String(5))

    player_seasons = relationship("PlayerSeason", back_populates='position')

