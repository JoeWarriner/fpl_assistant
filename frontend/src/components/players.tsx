import type {Player} from '../types/player';
import React from 'react';


export function PlayerRow(player: Player): JSX.Element{
    return (
            <tr>
                <td> {player.first_name} </td>
                <td> {player.second_name} </td>
                <td> {player.position} </td>
            </tr>
        
    );
    
}
