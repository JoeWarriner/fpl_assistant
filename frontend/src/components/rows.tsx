import type {Player} from '../types/player';
import React from 'react';


interface rowProps {
    first_name: string,
    second_name: string,
    position: string
}

export function PlayerRow(props: rowProps){
    return (
            <tr>
                <td> {props.first_name} </td>
                <td> {props.second_name} </td>
                <td> {props.position} </td>
            </tr>  
    );    
}

