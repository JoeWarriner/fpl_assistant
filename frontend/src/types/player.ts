
export type Player = {
    id: number;
    first_name: string;
    second_name: string;
    position: string;
}
