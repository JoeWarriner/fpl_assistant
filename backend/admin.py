from database.data_access_layer import dal
from etl.jobs.base_job import Job
from etl.imports.initial_import import InitialImport
from etl.imports.regular_import import RegularImport

from argparse import ArgumentParser

command_functions = {
    'update' : RegularImport,
    'init': InitialImport
}

parser = ArgumentParser()
parser.add_argument('command')
args = parser.parse_args()
task = command_functions.get(args.command)()
dal.connect()
if args.command == 'init':
    dal.reset_tables()
dal.session = dal.Session()
try:
    task.run()
except Exception as e:
    dal.session.rollback()
    raise Exception from e
else:
    dal.session.commit()
finally:
    dal.session.close()


