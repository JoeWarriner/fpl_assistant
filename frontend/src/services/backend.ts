import {Player} from '../types/player';


export type DataGetter  = (pageSize: number, offSet: number) => Promise<Player[]>


export const getPlayers = async (pageSize: number, offSet: number): Promise<Player[]> => {
    const baseUrl = "http://127.0.0.1:8000/players";
    const requestUrl = baseUrl.concat('?', 'pagesize=', pageSize.toString(), '&', 'offset=', offSet.toString())
    try {
        const response = await fetch(requestUrl);
        const team: Player[] = await response.json() as Player[];
        return team;
    } catch (error)  {
        console.log('Error getting data from server: ', error);
        return [];
    }
}


export const getOptimalTeam = async (pageSize: number, offSet: number) => {
    try {
        const response = await fetch("http://127.0.0.1:8000/optimised-team");
        const team: Player[] = await response.json() as Player[];
        return team;
    } catch (error)  {
        console.log('Error getting data from server: ', error);
        return [];
    }
}