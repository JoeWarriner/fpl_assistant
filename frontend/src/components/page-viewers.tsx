import type {Player} from '../types/player';
import {getPlayers} from  '../services/backend';
import {PlayerTable} from './tables'
import Button from 'react-bootstrap/Button';
import React, { useState, useEffect } from 'react';


export function PlayerPageViewer(){

    const [offSet, setOffSet] = useState<number>(0);

    const onClickPrevious = () => {
        setOffSet(offSet - 10);
    }
    const onClickNext = () => {
        setOffSet(offSet + 10);
    }

    return (
        <div>
            <PlayerTable 
                offSet={offSet} 
            />
            <Button
                onClick={onClickPrevious}
            >Previous</Button>
            <Button 
                onClick={onClickNext}
            >Next</Button>
        </div>
    )

        

}
