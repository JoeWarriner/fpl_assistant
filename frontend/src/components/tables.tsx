import type {Player} from '../types/player';
import {getPlayers} from  '../services/backend';
import {PlayerRow} from './rows'
import React, { useState, useEffect } from 'react';


interface PagePosition {
    offSet: number
}

export function PlayerTable(props: PagePosition){

    const [players, setPlayers] = useState<Player[]>([]);

    React.useEffect(() => {
        getPlayers(10, props.offSet).then((result) => {
            setPlayers(result);
        }).catch((error) => {
            console.log("Error fetching data", error);
        });
    },[props.offSet]);

    return (
        <table>
            <th>
                <th> First Name </th> 
                <th> Second Name </th> 
                <th> Position </th> 
            </th>
            {players.map(
                (player) => <PlayerRow
                    key={player.id} 
                    first_name={player.first_name} 
                    second_name={player.second_name}
                    position={player.position} />
 
            )
            }
        </table>
    );
}
