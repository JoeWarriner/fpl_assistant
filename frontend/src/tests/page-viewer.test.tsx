import renderer from 'react-test-renderer';
import React from 'react';
import {Player} from '../types/player';
import { PlayerTable } from '../components/tables';
import * as Backend from '../services/backend';
import { render } from 'react-dom';

jest.mock('../services/backend')


const getPlayersMock = async (pageSize: number, offSet: number): Promise<Player[]> => {
    const players: Player[] = await new Promise((resolve, reject) => {
            process.nextTick(() => {
                let testPlayer: Player
                if (offSet == 1){
                    testPlayer = {
                        id: 1,
                        first_name: 'Billy',
                        second_name: 'Bob',
                        position: 'DEF',
                        team: 'ARS',
                        predicted_score: 5,
                        current_value: 5
    
                    }
                } else {
                    testPlayer = {
                        id: 2,
                        first_name: 'Sammy',
                        second_name: 'Ray',
                        position: 'FWD',
                        team: 'ARS',
                        predicted_score: 5,
                        current_value: 5
                    }
                }
                testPlayer ?
                resolve([testPlayer]) :
                reject({error: 'Something went wrong'})
            })
            }
        )
        return players
    

}



it('player table offset 1', async () => {
    let table: renderer.ReactTestRenderer | undefined;
    

    await renderer.act(() => {
            table = renderer.create(<PlayerTable 
                                        offSet={1}
                                        pageSize={1}
                                        dataGetter={getPlayersMock} 
                                    />)
    })
   
    
    let tree = table?.toJSON()
    expect(tree).toMatchSnapshot();

    
    await renderer.act(() => {
        table = renderer.create(<PlayerTable 
                    offSet={2}
                    pageSize={1}
                    dataGetter={getPlayersMock} 
                />)
            })

    tree = table?.toJSON()
    expect(tree).toMatchSnapshot();



    })

