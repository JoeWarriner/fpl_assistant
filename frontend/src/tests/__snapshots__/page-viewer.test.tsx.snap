// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player table offset 1 1`] = `
<table>
  <th>
    <th>
       First Name 
    </th>
    <th>
       Second Name 
    </th>
    <th>
       Position 
    </th>
  </th>
  <tr>
    <td>
       
      Billy
       
    </td>
    <td>
       
      Bob
       
    </td>
    <td>
       
      DEF
       
    </td>
  </tr>
</table>
`;

exports[`player table offset 1 2`] = `
<table>
  <th>
    <th>
       First Name 
    </th>
    <th>
       Second Name 
    </th>
    <th>
       Position 
    </th>
  </th>
  <tr>
    <td>
       
      Sammy
       
    </td>
    <td>
       
      Ray
       
    </td>
    <td>
       
      FWD
       
    </td>
  </tr>
</table>
`;
